File: python/fm_index.py
from typing import List, Dict, Tuple

def suffix_array_slow(s: str) -> List[int]:
    """Solo para validar FM-Index si se necesita (no se usa en bench)."""
    return sorted(range(len(s)), key=lambda i: s[i:])

def bwt_from_sa(s: str, sa: List[int]) -> str:
    n = len(s)
    return "".join(s[i-1] if i != 0 else s[-1] for i in sa)

class FMIndex:
    """
    FM-Index con Occ muestreado para ahorrar memoria.
    - C[c]: número de caracteres < c
    - Occ(c, i): # de c en BWT[0:i]
    Muestreo cada 'step' entradas.
    """
    def __init__(self, s: str, sa: List[int], step: int = 128):
        assert s[-1] == "$", "El texto debe terminar con sentinela '$'."
        self.s = s
        self.sa = sa
        self.n = len(s)
        self.step = step

        self.bwt = bwt_from_sa(s, sa)
        # Alfabeto ordenado
        self.alpha = sorted(set(self.bwt))
        # C array
        counts = {c: 0 for c in self.alpha}
        for ch in self.bwt:
            counts[ch] += 1
        total = 0
        self.C = {}
        for c in sorted(self.alpha):
            self.C[c] = total
            total += counts[c]

        # Occ muestreado
        self.checkpoints: Dict[str, List[int]] = {c: [0] for c in self.alpha}
        running = {c: 0 for c in self.alpha}
        for i, ch in enumerate(self.bwt, 1):
            running[ch] += 1
            if i % self.step == 0:
                for c in self.alpha:
                    self.checkpoints[c].append(running[c])
        # Guardar cuentas finales para facilitar límite superior
        self.running_final = running

    def _occ(self, c: str, i: int) -> int:
        """# de c en BWT[0:i] (i exclusivo)."""
        if i <= 0:
            return 0
        block = i // self.step
        offset = i % self.step
        base = self.checkpoints[c][block] if block < len(self.checkpoints[c]) else self.running_final[c]
        # Escanear remainder
        start = block * self.step
        for j in range(start, start + offset):
            if j >= self.n: break
            if self.bwt[j] == c:
                base += 1
        return base

    def search(self, pattern: str) -> List[int]:
        """Retorna posiciones (en SA) de todas las ocurrencias de `pattern` por backward search."""
        if not pattern:
            return []
        l, r = 0, self.n
        for ch in reversed(pattern):
            if ch not in self.C:
                return []
            l = self.C[ch] + self._occ(ch, l)
            r = self.C[ch] + self._occ(ch, r)
            if l >= r:
                return []
        # mapear rango [l, r) a SA
        return [self.sa[i] for i in range(l, r)]

File: cpp/src/fm_index.cpp
#include "fm_index.hpp"
#include "suffix_array_mm.hpp"
#include <algorithm>
#include <cstring>

static std::string bwt_from_sa(const std::string& s, const std::vector<int>& sa) {
    int n = (int)s.size();
    std::string bwt(n, '\0');
    for (int i = 0; i < n; ++i) {
        int idx = sa[i];
        bwt[i] = (idx == 0) ? s[n-1] : s[idx-1];
    }
    return bwt;
}

FMIndex::FMIndex(const std::string& s, const std::vector<int>& sa_, int step_)
: sa(sa_), step(step_), n((int)s.size()) {
    bwt = bwt_from_sa(s, sa);
    C.assign(256, 0);
    std::vector<int> cnt(256, 0);
    for (char ch : bwt) cnt[(unsigned char)ch]++;
    int tot = 0;
    for (int c = 0; c < 256; ++c) {
        C[c] = tot;
        tot += cnt[c];
    }
    int blocks = (n + step - 1) / step;
    checkpoints.assign(256, std::vector<int>(blocks+1, 0));
    std::vector<int> run(256, 0);
    for (int i = 1; i <= n; ++i) {
        unsigned char ch = (unsigned char)bwt[i-1];
        run[ch]++;
        if (i % step == 0) {
            int b = i / step;
            for (int c = 0; c < 256; ++c) checkpoints[c][b] = run[c];
        }
    }
    // almacenar totales al final en checkpoints[][blocks]
    for (int c = 0; c < 256; ++c) checkpoints[c][blocks] = run[c];
}

int FMIndex::occ(unsigned char c, int i) const {
    if (i <= 0) return 0;
    int block = i / step;
    int offset = i % step;
    int base = checkpoints[c][block];
    int start = block * step;
    int end = std::min(start + offset, n);
    for (int j = start; j < end; ++j) {
        if ((unsigned char)bwt[j] == c) base++;
    }
    return base;
}

std::vector<int> FMIndex::search(const std::string& pat) const {
    if (pat.empty()) return {};
    int l = 0, r = n;
    for (int k = (int)pat.size()-1; k >= 0; --k) {
        unsigned char c = (unsigned char)pat[k];
        int cbase = C[c];
        l = cbase + occ(c, l);
        r = cbase + occ(c, r);
        if (l >= r) return {};
    }
    std::vector<int> out;
    out.reserve(r - l);
    for (int i = l; i < r; ++i) out.push_back(sa[i]);
    return out;
}
